cluster-autoscaler/expander/priority/priority.go

```diff
@@ -0,0 +1,189 @@
+				res.logConfigWarning("PriorityConfigMapInvalid", msg)
+				continue
+			}
+		}
+	}()
+	return res, nil
+}
+
+func (p *priority) logConfigWarning(reason, msg string) {
+	p.logRecorder.Event(apiv1.EventTypeWarning, reason, msg)
+	klog.Warning(msg)
+	p.badConfigUpdates++
+}
+
+func (p *priority) parsePrioritiesYAMLString(prioritiesYAML string) error {
+	if prioritiesYAML == "" {
+		p.badConfigUpdates++
+		return fmt.Errorf("priority configuration in %s configmap is empty; please provide valid configuration", PriorityConfigMapName)
+	}
+	var config map[int][]string
+	if err := yaml.Unmarshal([]byte(prioritiesYAML), &config); err != nil {
+		p.badConfigUpdates++
+		return fmt.Errorf("Can't parse YAML with priorities in the configmap: %v", err)
+	}
+
+	newPriorities := make(map[int][]*regexp.Regexp)
+	for prio, reList := range config {
+		for _, re := range reList {
+			regexp, err := regexp.Compile(re)
+			if err != nil {
+				p.badConfigUpdates++
+				return fmt.Errorf("Can't compile regexp rule for priority %d and rule %s: %v", prio, re, err)
+			}
+			newPriorities[prio] = append(newPriorities[prio], regexp)
+		}
+	}
+
+	p.padlock.Lock()
+	p.priorities = newPriorities
+	p.okConfigUpdates++
+	p.padlock.Unlock()
+
+	msg := "Successfully reloaded priority configuration from configmap."
+	klog.V(4).Info(msg)
+	p.logRecorder.Event(apiv1.EventTypeNormal, "PriorityConfigMapReloaded", msg)
+
+	return nil
+}
+
+func (p *priority) BestOption(expansionOptions []expander.Option, nodeInfo map[string]*schedulernodeinfo.NodeInfo) *expander.Option {
+	if len(expansionOptions) <= 0 {
+		return nil
+	}
+
+	maxPrio := -1
+	best := []expander.Option{}
+	p.padlock.RLock()
+	for _, option := range expansionOptions {
+		id := option.NodeGroup.Id()
+		found := false
+		for prio, nameRegexpList := range p.priorities {
+			if prio < maxPrio {
+				continue
+			}
+			if !p.groupIDMatchesList(id, nameRegexpList) {
+				continue
+			}
+			if prio > maxPrio {
+				maxPrio = prio
+				best = nil
+			}
+			best = append(best, option)
+			found = true
+			break
+		}
+		if !found {
+			msg := fmt.Sprintf("Priority expander: node group %s not found in priority expander configuration. "+
+				"The group won't be used.", id)
+			p.logConfigWarning("PriorityConfigMapNotMatchedGroup", msg)
+		}
+	}
+	p.padlock.RUnlock()
+
+	if len(best) == 0 {
+		msg := "Priority expander: no priorities info found for any of the expansion options. Falling back to random choice."
+		p.logConfigWarning("PriorityConfigMapNoGroupMatched", msg)
+		return p.fallbackStrategy.BestOption(expansionOptions, nodeInfo)
+	}
+
+	return p.fallbackStrategy.BestOption(best, nodeInfo)
+}
+
+func (p *priority) groupIDMatchesList(id string, nameRegexpList []*regexp.Regexp) bool {
+	for _, re := range nameRegexpList {
+		if re.FindStringIndex(id) != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// EventRecorder is an interface to abstract kubernetes event recording.
+type EventRecorder interface {
```

Comment:
Why are you redefining this here?

Reply:
For testability - check unit tests. Have I missed exactly the same interface being already defined elsewhere?

Reply:
After taking a closer look at unittests - they're very inconsistent with existing unittests. We have utilities we use for making fake nodegroups (https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/test/test_cloud_provider.go) and there is an existing utility for faking event recording (used for example here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/processors/status/eventing_scale_up_processor_test.go#L87).

Please use existing tooling for your unittests.